test/integration/select_test.lua

```diff
@@ -219,6 +219,126 @@ add('test_select_all_with_limit', function(g)
     t.assert_equals(#objects, 0)
 end)
 
+add('test_negative_first', function(g)
+    local customers = insert_customers(g, {
+        {
+            id = 1, name = "Elizabeth", last_name = "Jackson",
+            age = 12, city = "New York",
+        }, {
+            id = 2, name = "Mary", last_name = "Brown",
+            age = 46, city = "Los Angeles",
+        }, {
+            id = 3, name = "David", last_name = "Smith",
+            age = 33, city = "Los Angeles",
+        }, {
+            id = 4, name = "William", last_name = "White",
+            age = 81, city = "Chicago",
+        }, {
+            id = 5, name = "Jack", last_name = "Sparrow",
+            age = 35, city = "London",
+        }, {
+            id = 6, name = "William", last_name = "Terner",
+            age = 25, city = "Oxford",
+        }, {
+            id = 7, name = "Elizabeth", last_name = "Swan",
+            age = 18, city = "Cambridge",
+        }, {
+            id = 8, name = "Hector", last_name = "Barbossa",
+            age = 45, city = "London",
+        },
+    })
+
+    table.sort(customers, function(obj1, obj2) return obj1.id < obj2.id end)
+
+    -- negative first w/o after
+    local first = -10
+    local objects, err = g.cluster.main_server.net_box:eval([[
+        local crud = require('crud')
+
+        local first = ...
+
+        local objects, err = crud.select('customers', nil, {
+            first = first,
+        })
+        return objects, err
+    ]], {first})
+
+    t.assert_equals(objects, nil)
+    t.assert_str_contains(err.err, "Negative first should be specified only with after option")
+
+    -- no conditions
+    -- first -3 after 5 (batch_size is 1)
+    local first = -3
+    local after = customers[5]
+    local batch_size = 1
+    local objects, err = g.cluster.main_server.net_box:eval([[
+        local crud = require('crud')
+
+        local first, after, batch_size = ...
+
+        local objects, err = crud.select('customers', nil, {
+            first = first,
+            after = after,
+            batch_size = batch_size,
+        })
+        return objects, err
+    ]], {first, after, batch_size})
+
+    t.assert_equals(err, nil)
+    t.assert_equals(objects, get_by_ids(customers, {2, 3, 4}))
+
+    -- age <= 33 (eq for id: 3)
+    -- (in sorted by age order: 4, 2, 8, 5, 3, 6, 7, 1)
```

Comment:
I hope you perfectly understand that this test shouldn't test that you query planner works fine.
And such comments and conditions just overcomplicate test logic.

I suggest you to use "id > 100" with reverse pagination - it allows to preserve clear logic (all your elements will be placed in order that specified at start of this test, by id). Some person more easily understand this logic and you don't make resort the data in mind or verify that specified result here is correct.

Also you need to check several basic cases:
  - you get inverted result set
  - you invert them at router not storage

To test first case you just need to query all data. To test the second case you need to query all records one by one.

Reply:
I've added this check for `age` condition to be sure that scan value is set properly for this case.

Reply:
Well, test looks good now. Thanks.